src/System.Management.Automation/namespaces/FileSystemContentStream.cs

```diff
@@ -287,13 +288,34 @@ internal class FileSystemContentReaderWriter : IContentReader, IContentWriter
                 // For Boyer-Moore string search algorithm.
                 // Populate the offset lookups.
                 // These will tell us the maximum number of characters
-                // we can read to generate another possible match.
+                // we can read to generate another possible match (safe shift).
                 // If we read more characters than this, we risk consuming
                 // more of the stream than we need.
-                _offsetDictionary = new Dictionary<char, int>(_delimiter.Length);
-                foreach (char currentChar in _delimiter)
+                //
+                // Because an unicode character size is 2 byte we would to have use
+                // very large array with 65535 size to keep this safe offsets.
+                // One solution is to pack unicode character to byte.
+                // The workaround is to use low byte from unicode character.
+                // This allow us to use small array with size 256.
+                // This workaround is the fastest and provides excellent results
+                // in regular search scenarios when the file contains
+                // mostly characters from the same alphabet.
+                _offsetDictionary = new int[256];
+
+                // If next char from file is not in search pattern safe shift is the search pattern length.
+                for (var n = 0 ; n < _offsetDictionary.Length; n++)
                 {
-                    _offsetDictionary[currentChar] = _delimiter.Length - _delimiter.LastIndexOf(currentChar) - 1;
+                    _offsetDictionary[n] = _delimiter.Length;
+                }
+
+                // If next char from file is in search pattern we should calculate a safe shift.
+                char currentChar;
+                byte lowByte;
+                for (var i = 0 ; i < _delimiter.Length; i++)
+                {
+                    currentChar = _delimiter[i];
+                    lowByte = Unsafe.As<char,byte>(ref currentChar);
+                    _offsetDictionary[lowByte] = _delimiter.Length - _delimiter.LastIndexOf(currentChar) - 1;
```

Comment:
I think this can be
```
_offsetDictionary[lowByte] = _delimiter.Length - i - 1;
```
For duplicate chars the last index will end up in _offsetDictionary[lowByte]

Also, do we need to be worried about char collisions using the low byte?

Reply:
> Also, do we need to be worried about char collisions using the low byte?

There is a comment in the code. If we have a file with chars from many languages we'll have many collisions and lost performance. In practice, we can expect that files contains chars from one-two languages. In the common case we'll have no collisions and best performance.

Reply:
> I think this can be

Thanks! Good catch!